### PR TITLE
gh-121528: Fix the process crash when load extension when it's compiled with low version of the python.

### LINKED ISSUE #121528: [3.13] _PyObject_Init() assertion fails with the stable ABI on Python built with assertions (extension built with Python 3.11)
# Crash report

### What happened?

Consider the following extension.

C code:

```c
#include <Python.h>

static PyObject *
foo_bar(PyObject *self, PyObject *args)
{
	Py_INCREF(PyExc_TypeError);
	PyErr_SetString(PyExc_TypeError, "foo");
	return NULL;
}

static PyMethodDef foomethods[] = {
	{"bar", foo_bar, METH_VARARGS, ""},
	{NULL, NULL, 0, NULL},
};

static PyModuleDef foomodule = {
	PyModuleDef_HEAD_INIT,
	.m_name = "foo",
	.m_doc = "foo test module",
	.m_size = -1,
	.m_methods = foomethods,
};

PyMODINIT_FUNC
PyInit_foo(void)
{
	return PyModule_Create(&foomodule);
}
```

`setup.py`:

```python
from setuptools import setup, Extension

setup(name='foo',
      version='0',
      ext_modules=[
          Extension('foo', ['foo.c'], py_limited_api='cp38'),
      ])
```

If I compile the extension using Python older than 3.12, and then run the method, Python 3.13.0b3 (built `--with-assertions`) crashes:

```
$ python3.11 setup.py build_ext -i
running build_ext
building 'foo' extension
creating build
creating build/temp.linux-x86_64-cpython-311
x86_64-pc-linux-gnu-gcc -Wsign-compare -fPIC -I/usr/include/python3.11 -c foo.c -o build/temp.linux-x86_64-cpython-311/foo.o
creating build/lib.linux-x86_64-cpython-311
x86_64-pc-linux-gnu-gcc -shared build/temp.linux-x86_64-cpython-311/foo.o -L/usr/lib64 -o build/lib.linux-x86_64-cpython-311/foo.abi3.so
copying build/lib.linux-x86_64-cpython-311/foo.abi3.so -> 
$ python3.13 -c 'import foo; foo.bar()'
python3.13: ./Include/internal/pycore_object.h:284: _PyObject_Init: Assertion `_PyType_HasFeature(typeobj, Py_TPFLAGS_HEAPTYPE) || _Py_IsImmortal(typeobj)' failed.
Aborted (core dumped)
```

I've been able to bisect it to c32dc47aca6e8fac152699bc613e015c44ccdba9 (CC @markshannon). Originally hit it in extensions using PyO3, and reported to https://github.com/PyO3/pyo3/issues/4311.

### CPython versions tested on:

CPython main branch

### Operating systems tested on:

Linux

### Output from running 'python -VV' on the command line:

Python 3.13.0b3 (main, Jul  4 2024, 14:30:57) [GCC 14.1.1 20240622]

### Linked PRs
* gh-121550
* gh-121725
* gh-121936

### PR DESCRIPTION
**Pull request by Zheaoli**
Fix #121528 

* Issue: gh-121528

### CODE REVIEW

File: Include/internal/pycore_object.h [hunk 1]
```diff
@@ -266,8 +266,9 @@ _PyObject_Init(PyObject *op, PyTypeObject *typeobj)
 {
     assert(op != NULL);
     Py_SET_TYPE(op, typeobj);
-    assert(_PyType_HasFeature(typeobj, Py_TPFLAGS_HEAPTYPE) || _Py_IsImmortal(typeobj));
-    Py_INCREF(typeobj);
+    if (_PyType_HasFeature(typeobj, Py_TPFLAGS_HEAPTYPE) || _Py_IsImmortal(typeobj)){
```
Comment (ZeroIntensity) [inline Include/internal/pycore_object.h]: This should follow [PEP 7](https://peps.python.org/pep-0007/). There should be a space between the end of the parens and `{`

**Review by ZeroIntensity — CHANGES_REQUESTED** [review]
Can confirm that this PR fixes #121528

**Comment by ZeroIntensity** [comment]
@Eclips4 Can you add the 3.13 backport label?

**Comment by Eclips4** [comment]
cc @mgorny can you confirm that this patch fixes the issue?

File: Misc/NEWS.d/next/Core_and_Builtins/2024-07-10-03-13-19.gh-issue-121528._Bo7rw.rst [hunk 2]
```diff
@@ -0,0 +1,2 @@
+Fix the process crash when load extension when it's compiled with low
```
Comment (ZeroIntensity) [inline Misc/NEWS.d/next/Core_and_Builtins/2024-07-10-03-13-19.gh-issue-121528._Bo7rw.rst]: I think this should be reworded, it's not too clear to someone who's not familiar with the exact issue.

```suggestion
Fix crash when loading extension modules that are compiled with an older
```

File: Misc/NEWS.d/next/Core_and_Builtins/2024-07-10-03-13-19.gh-issue-121528._Bo7rw.rst [hunk 3]
```diff
@@ -0,0 +1,2 @@
...
+version of the python.
```
Comment (ZeroIntensity) [inline Misc/NEWS.d/next/Core_and_Builtins/2024-07-10-03-13-19.gh-issue-121528._Bo7rw.rst]:
```suggestion
version of Python.
```

**Review by ZeroIntensity — CHANGES_REQUESTED** [review]

**Comment by Zheaoli** [comment]
@ZeroIntensity Thanks for the tips!

**Review by ZeroIntensity — APPROVED** [review]
Thanks for the changes, LGTM.

**Pull request closed without merging**